File: Search_Sorted_Array_of_Unknown_Size.py
# Time Complexity : O(logN)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : A little unsure about the Precise Time Complexity
# since we have to search for low and high, would that affect the time complexity

class Solution:
    def search(self, reader: 'ArrayReader', target: int) -> int:
        low:int = 0
        high:int = 1
        # defining the search space to conduct binary seach
        while target > reader.get(high):
            low = high + 1
            high *= 2

        while low <= high:
            mid:int = low + ((high-low)//2)

            if reader.get(mid) > target:
                high = mid - 1
            elif reader.get(mid) < target:
                low = mid + 1
            else:
                return mid
        return -1

File: SearchSortedArrayOfUnknownSize.cpp
// Time Complexity : O(logn)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :


/**
 * // This is the ArrayReader's API interface.
 * // You should not implement it, or speculate about its implementation
 * class ArrayReader {
 *   public:
 *     int get(int index);
 * };
 */

class Solution {
public:
    int search(const ArrayReader& reader, int target) {
        int low = 0; 
        int high = findLastIndex(reader);

        while(low <= high) {
            int mid = (low+high)/2; 
            if(reader.get(mid) == target) return mid; 
            if(reader.get(mid) < target) low = mid+1; 
            else high = mid-1; 
        }

        return -1; 
    }

    int findLastIndex(const ArrayReader& reader) {
        int low = 0; 
        int high = 10000; 

        while(low <= high) {
            int mid = (low+high)/2;
            long long midElement = reader.get(mid);
            if(midElement != INT_MAX) {
                if(reader.get(mid+1) == INT_MAX) return mid; 
                low = mid+1; 
            } else {
                high = mid-1;
            }
        }

        return -1; 
    }
};